test/docker-e2e/e2e_block_sync_test.go

```diff
@@ -0,0 +1,109 @@
+	}
+
+	ctx := context.TODO()
+	cfg := dockerchain.DefaultConfig(s.client, s.network)
+	celestia, err := dockerchain.NewCelestiaChainBuilder(s.T(), cfg).Build(ctx)
+	s.Require().NoError(err, "failed to create chain")
+
+	t.Cleanup(func() {
+		if err := celestia.Stop(ctx); err != nil {
+			t.Logf("Error stopping chain: %v", err)
+		}
+	})
+
+	err = celestia.Start(ctx)
+	s.Require().NoError(err, "failed to start chain")
+
+	// verify the chain is running
+	height, err := celestia.Height(ctx)
+	s.Require().NoError(err, "failed to get chain height")
+	s.Require().Greater(height, int64(0), "chain height is zero")
+
+	s.CreateTxSim(ctx, celestia)
+
+	allNodes := celestia.GetNodes()
+
+	nodeClient, err := allNodes[0].GetRPCClient()
+	s.Require().NoError(err)
+
+	initialHeight, err := s.GetLatestBlockHeight(ctx, nodeClient)
+	s.Require().NoError(err, "failed to get initial height")
+	s.Require().Greater(initialHeight, int64(0), "initial height is zero")
+
+	targetHeight := initialHeight + blockSyncBlocksToProduce
+	t.Logf("Successfully reached initial height %d", initialHeight)
+
+	s.Require().NoError(wait.ForBlocks(ctx, int(targetHeight), celestia), "failed to wait for target height")
+
+	t.Logf("Successfully reached target height %d", targetHeight)
+
+	// get the latest height and block info for the block sync node
+	latestHeight, err := s.GetLatestBlockHeight(ctx, nodeClient)
+	s.Require().NoError(err, "failed to get latest height")
+	s.Require().Greater(latestHeight, int64(0), "latest height is zero")
+
+	// build peer list for the new node to connect to existing validators
+	peerList, err := addressutil.BuildInternalPeerAddressList(ctx, celestia.GetNodes())
+
+	t.Logf("Latest height: %d", latestHeight)
+	t.Logf("Peers: %s", peerList)
+
+	t.Log("Adding block sync node")
+	err = celestia.AddNode(ctx,
+		celestiadockertypes.NewChainNodeConfigBuilder().
+			WithNodeType(celestiadockertypes.FullNodeType).
+			WithPostInit(func(ctx context.Context, node *celestiadockertypes.ChainNode) error {
+				return config.Modify(ctx, node, "config/config.toml", func(cfg *cometcfg.Config) {
+					// disable state sync to ensure we're testing block sync
+					cfg.StateSync.Enable = false
+					// configure block sync
+					cfg.BlockSync.Version = "v0"
+					// set persistent peers to connect to existing nodes
+					cfg.P2P.PersistentPeers = peerList
+				})
+			}).
+			Build(),
+	)
+
+	s.Require().NoError(err, "failed to add block sync node")
+
+	allNodes = celestia.GetNodes()
+	blockSyncNode := allNodes[len(allNodes)-1]
+
+	s.Require().Equal("fn", blockSyncNode.GetType(), "expected block sync node to be a full node")
+
+	blockSyncClient, err := blockSyncNode.GetRPCClient()
+	s.Require().NoError(err)
+
+	err = s.WaitForSync(ctx, blockSyncClient, blockSyncTimeout, func(info rpctypes.SyncInfo) bool {
+		return !info.CatchingUp && info.LatestBlockHeight >= latestHeight
+	})
+
+	s.Require().NoError(err, "failed to wait for block sync node to catch up")
```

Comment:
[question] in a follow-up test or in this test, would it make sense for the chain to have a few blocks on app version 1, 2, 3, and 4?

I think it is uncommon to hit regressions with block syncing if all the blocks are on the latest version and the block sync'ing node starts on the latest version.

However, I think this test may have greater coverage and help us catch future regressions if it block synced through blocks with all historical app versions.

Reply:
Yes I think we can maybe have  a more elaborate test in which we perform incremental upgrades and make sure we have blocks at each version. The upgrade logic is being handled as its own test in https://github.com/celestiaorg/celestia-app/pull/5113 when that is merged we can make a hybrid of the two, and potentially remove this one or just run it less frequently if there aren't regressions with same version block syncs that often

Reply:
https://github.com/celestiaorg/celestia-app/issues/5261